### PR TITLE
Avoid marshaling EVE-JSON with fixed schema when modifying items

### PR DESCRIPTION
**Pull request by satta**
This PR addresses #54 by avoiding the use of JSON marshaling when handling EVE-JSON received from Suricata. Consistent use of marshaling would require that a full schema of the Suricata version that is deployed is known to and updated in FEVER at all times, otherwise one would risk dropping data that is in the input but not present in the schema.
The proposed new implementation always uses `buger/jsonparser`'s `Set` function to add or modify fields in the original EVE-JSON, for example when creating an alert from a metadata event. Tests still finish successfully.

To check for the correctness of these modifications, I have also added unit tests for the reverse DNS enricher, which have been missing so far. To do that I had to refactor `HostNamer` into an interface to allow more flexible dependency injection, making it possible to use a mock `HostNamer` to test properly in situations where one cannot rely on proper Internet access.
In addition, I addressed some `golint` issues about constant naming and fixed a race condition caused by a debug output call.

### CODE REVIEW

File: processing/stenosis_connector_aux_test.go [hunk 1]
```diff
@@ -30,11 +30,11 @@ type mockGRPCServerOptionFunc func(*mockGRPCServer)
 
 type mockGRPCServerTokenGenFunc func(*task.Query) string
 
-func mockGRPCServerFailOption(meth string, err error) mockGRPCServerOptionFunc {
-	return func(m *mockGRPCServer) {
-		m.failWith[meth] = err
-	}
-}
+//func mockGRPCServerFailOption(meth string, err error) mockGRPCServerOptionFunc {
```
Comment (geertjanvdk) [inline processing/stenosis_connector_aux_test.go:33]: Always good to comment code that is being commented out. Otherwise it looks like forgotten debugging code, or dead code.

File: processing/bloom_handler.go [hunk 2]
```diff
@@ -13,78 +12,82 @@ import (
...
 // indicator match, e.g. a Bloom filter hit. The resulting alert will retain
 // the triggering event's metadata (e.g. 'dns' or 'http' objects) as well as
 // its timestamp.
 func MakeAlertEntryForHit(e types.Entry, eType string, alertPrefix string, ioc string) types.Entry {
-	var eve types.EveEvent
-	var newEve types.EveEvent
-	var err = json.Unmarshal([]byte(e.JSONLine), &eve)
+
+	var value string
+	if eType == "http-url" {
```
Comment (geertjanvdk) [inline processing/bloom_handler.go:36]: Nothing wrong, but maybe a switch is easier on the eyes :)

```go
switch {
case eType == ...:
case strings.HasPrrefix...:
}
```

File: processing/bloom_handler.go [hunk 3]
```diff
@@ -13,78 +12,82 @@ import (
...
+	} else if eType == "tls-sni" {
+		value = e.TLSSni
+	}
+	var sig = "%s Possibly bad traffic: %s"
+	if v, ok := sigs[eType]; ok {
+		sig = v
+	}
+	newEntry := e
+	newEntry.EventType = "alert"
+	l, err := jsonparser.Set([]byte(newEntry.JSONLine), []byte("\"alert\""), "event_type")
```
Comment (geertjanvdk) [inline processing/bloom_handler.go:51]: Avoid escaping quotes, use backticks instead.

File: processing/bloom_handler.go [hunk 4]
```diff
@@ -13,78 +12,82 @@ import (
...
-			PacketInfo: eve.PacketInfo,
-			HTTP:       eve.HTTP,
-			DNS:        eve.DNS,
-			TLS:        eve.TLS,
-			ExtraInfo: &types.ExtraInfo{
-				BloomIOC: ioc,
-			},
+		newEntry.JSONLine = string(l)
+	}
+	l, err = jsonparser.Set([]byte(newEntry.JSONLine), []byte("\"allowed\""), "alert", "action")
```
Comment (geertjanvdk) [inline processing/bloom_handler.go:57]: Would be good to have some newlines here and there. Hard to follow IF-statements.

File: processing/bloom_handler.go [hunk 5]
```diff
@@ -13,78 +12,82 @@ import (
...
+	if err != nil {
+		log.Warning(err)
+	} else {
+		newEntry.JSONLine = string(l)
+	}
+	l, err = jsonparser.Set([]byte(newEntry.JSONLine), []byte("\"Potentially Bad Traffic\""), "alert", "category")
```
Comment (geertjanvdk) [inline processing/bloom_handler.go:63]: It looks like the code is not using the `if l, err := ...; err != nil` Go thingy? It would make sense here, since `l` is then within the scope of the if statement only?

File: processing/bloom_handler_test.go [hunk 6]
```diff
@@ -46,6 +46,9 @@ func makeBloomDNSEvent(rrname string) types.Entry {
 		DNSType:   []string{"answer", "query"}[rand.Intn(2)],
 	}
 	eve := types.EveEvent{
+		Timestamp: &types.SuriTime{
+			Time: time.Now(),
```
Comment (geertjanvdk) [inline processing/bloom_handler_test.go:50]: Not `time.UTCNow()`?

File: processing/bloom_handler_test.go [hunk 7]
```diff
@@ -83,6 +86,9 @@ func makeBloomHTTPEvent(host string, url string) types.Entry {
 		HTTPMethod: "GET",
 	}
 	eve := types.EveEvent{
+		Timestamp: &types.SuriTime{
+			Time: time.Now(),
```
Comment (geertjanvdk) [inline processing/bloom_handler_test.go:90]: Not `time.UTCNow()`?

File: processing/bloom_handler_test.go [hunk 8]
```diff
@@ -115,6 +121,9 @@ func makeBloomTLSEvent(host string) types.Entry {
 		TLSSni:    host,
 	}
 	eve := types.EveEvent{
+		Timestamp: &types.SuriTime{
+			Time: time.Now(),
```
Comment (geertjanvdk) [inline processing/bloom_handler_test.go:125]: Not `time.UTCNow()`?

File: processing/ip_handler_test.go [hunk 9]
```diff
@@ -40,6 +40,9 @@ func makeIPHTTPEvent(srcip string, dstip string) types.Entry {
 		HTTPMethod: "GET",
 	}
 	eve := types.EveEvent{
+		Timestamp: &types.SuriTime{
+			Time: time.Now(),
```
Comment (geertjanvdk) [inline processing/ip_handler_test.go:44]: Yes, might not be important, but I rather ask anyway as this is all going to some kind of data store :)
Not `UTCNow()`?

File: processing/ip_handler_test.go [hunk 10]
```diff
@@ -136,6 +138,16 @@ func TestIPHandler(t *testing.T) {
 		t.Fatalf("expected %d forwarded BLF alerts, seen less (%d)", 2,
 			len(fwhandler.Entries))
 	}
+
+	var i interface{}
```
Comment (geertjanvdk) [inline processing/ip_handler_test.go:142]: Nitpicking, and +20 years of coding, but `i` is like bound to be an integer in any language ;)
Maybe call this the `result`? (I know why `i` is used, like `s` is for string..) Like I said, nitpicking!

File: util/hostnamer_rdns.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+	"time"
+
+	"github.com/patrickmn/go-cache"
+)
+
+// HostNamerRDNS is a component that provides cached hostnames for IP
+// addresses passed as strings, determined via reverse DNS lookups.
+type HostNamerRDNS struct {
+	Cache *cache.Cache
+	Lock  sync.Mutex
```
Comment (geertjanvdk) [inline util/hostnamer_rdns.go:16]: No need to export a Mutex, or?

File: processing/stenosis_connector_aux_test.go (hunk 1 above)
Comment (satta) [reply processing/stenosis_connector_aux_test.go:33]: It is indeed dead code, kept around because I do not know Remo's intention for it. Removed it as it can be reproduced from Git history if necessary.

File: processing/bloom_handler.go (hunk 2 above)
Comment (satta) [reply processing/bloom_handler.go:36]: True. Done.

File: processing/bloom_handler.go (hunk 3 above)
Comment (satta) [reply processing/bloom_handler.go:51]: Done, also in other code files.

File: processing/bloom_handler.go (hunk 4 above)
Comment (satta) [reply processing/bloom_handler.go:57]: Got it. Done.

File: processing/bloom_handler_test.go (hunk 6 above)
Comment (satta) [reply processing/bloom_handler_test.go:50]: This is code that just generates mock test input that would otherwise come from Suricata. The code being tested does not make use of the timestamp at all. It is only generated to have _some_ value in there to get a more realistic entry example. 
That being said, there's nothing to say against forcing UTC here. But I assume you meant ` time.Now().UTC()`, there is no `time.UTCNow()`?

File: processing/bloom_handler_test.go (hunk 7 above)
Comment (satta) [reply processing/bloom_handler_test.go:90]: See above.

File: processing/bloom_handler_test.go (hunk 8 above)
Comment (satta) [reply processing/bloom_handler_test.go:125]: See above.

File: processing/ip_handler_test.go (hunk 9 above)
Comment (satta) [reply processing/ip_handler_test.go:44]: None of this is going to a data store. It's just generating mock test data that would come from an outside input in a real case.

File: processing/ip_handler_test.go (hunk 10 above)
Comment (satta) [reply processing/ip_handler_test.go:142]: Good point.

File: util/hostnamer_rdns.go (hunk 11 above)
Comment (satta) [reply util/hostnamer_rdns.go:16]: Indeed none of these need to be exported, good catch.

File: processing/bloom_handler.go (hunk 5 above)
Comment (satta) [reply processing/bloom_handler.go:63]: Indeed. Changed.

**Comment by satta** [comment]
Thanks for the helpful review!
